### DISCUSSION TITLE
mep-114 Deprecation path

### DISCUSSION THREAD

**Discussion by davekeeshan** [myhdl/myhdl#392] [Ideas]
While update the github actions flow, I noticed some warnings (`UserWarnings` actually) floating around as the related to [mep-114](https://dev.myhdl.org/meps/mep-114.html).  This, as an update, is nearly 7 years old, and probably should be removed from the code base.

However as I looked through the pytests, I found a lot of the warnings strewn through out the code, and there were a lot of tests that were still doing it the old way.

`pytest` has a flag for escalating warnings to errors so that they will fail tests and you can find them. (`-W error::DeprecationWarning`) `UserWarning` is used other places too which are legit, and there are many of those. 

I think/suggest that all the `UserWarnings` connected to mep-114 should be escalated to `DeprecationWarnings`. I also think that testing should update all code that runs deprecated code, testing is supposed to represent your codebase present and future not past (if possible), though there should be some tests to check that the deprecation warnings are asserted when the deprecated condition is met, but that is all that test does.

This is a lot of work.  Don't worry I have done most it it already on a branch when I was trying to scope out this issue.

I also found out that there are some test constructs that pytest no longer support, as a result some tests were no longer being run.  These were also being reported via a warning, but whizz by, so in addition to escalting DeprecationWarnings to errors for test, I also thing that all pytest warning should be esclated to errors so that they can be caught, appropriatly, if an when pytest changes. `-W error::pytest.PytestWarning`

Please have a look at this action which fails due to a `DeprecationWarning.`

https://github.com/daxzio/myhdl/actions/runs/3656635296/jobs/6179301608

**Comment by josyb** [comment]
"A warning is an error waiting to happen"

**Reply by davekeeshan** [reply]
Totally agree, so we should escalate all  warnings to errors! ;)

Sadly it is a little more complicated here, some of the warnings are the tool reporting warnings, which is it's job so arguably it is doing it job correctly.  These however are hard to distinguish between proper warnings that should be addressed. This is further complicated when they all have the same report level, excpetion level so it can be hard to sort them out.

This bizarrely remoinds me too much of how most EDA tools report so many warnings and it isn't too long before there are so many people just begin to ignore them, and that is where the real problems start.

Any way, I take it from your comment you agree with the suggested change at least @josyb
